api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -513,29 +517,48 @@ private FirecloudWorkspaceId getFcBillingProject(DbAccessTier accessTier, Worksp
     }
     final ImmutableMap<String, WorkspaceAccessLevel> aclsByEmail = shareRolesMapBuilder.build();
 
+    List<String> usersHavePermission = new ArrayList<>();
+    // Revoke lifesciene permission before asking Firecloud to remove users. Because after unshare
+    // in
+    // Firecloud, we nolonger can get user petSA from SAM using their credential.
+    if (dbWorkspace.getCdrVersion().getAccessTier().getEnableUserWorkflows()) {
+      usersHavePermission =
+          aclsByEmail.entrySet().stream()
+              .filter(entry -> shouldGrantWorkflowRunnerAsService(userProvider.get(), entry))
+              .map(Map.Entry::getKey)
+              .collect(Collectors.toList());
+
+      // Find the users who are owners or writers before share, but not in the new gain permission
+      // list
+      List<String> finalUsersHavePermission = usersHavePermission;
```

Comment:
This variable is exclusively used to check presence, please use a Set instead of list